flang/lib/Lower/CallInterface.cpp

```diff
@@ -461,13 +461,37 @@ class Fortran::lower::CallInterfaceImpl {
     }
   }
 
+  // Some dummy argument attributes prevent it to be used with implicit
+  // interfaces but that odes not imply that a box is needed for those. For
+  // instance, scalar intrinsic that are OPTIONAL require an explicit interface,
+  // but they can still be passed by reference.
+  bool dummyRequiresBox(
+      const Fortran::evaluate::characteristics::DummyDataObject &obj) {
+    using ShapeAttr = Fortran::evaluate::characteristics::TypeAndShape::Attr;
+    using ShapeAttrs = Fortran::evaluate::characteristics::TypeAndShape::Attrs;
+    constexpr ShapeAttrs shapeRequiringBox = {
+        ShapeAttr::AssumedShape, ShapeAttr::DeferredShape,
+        ShapeAttr::AssumedRank, ShapeAttr::Coarray};
+    if ((obj.type.attrs() & shapeRequiringBox).any())
+      // Need to pass shape/coshape info in fir.box.
+      return true;
+    else if (obj.type.type().IsPolymorphic())
+      // Need to pass dynamic type info in fir.box.
+      return true;
+    else if (const auto *derived{
+                 Fortran::evaluate::GetDerivedTypeSpec(obj.type.type())})
+      // Need to pass type parameters in fir.box if any.
+      return derived->parameters().empty();
+    return false;
+  }
+
   void handleExplicitDummy(
       const Fortran::evaluate::characteristics::DummyDataObject &obj,
       const FortranEntity &entity) {
     using Attrs = Fortran::evaluate::characteristics::DummyDataObject::Attr;
 
-    if (obj.attrs.test(Attrs::Optional))
-      TODO("Optional in procedure interface");
+    // if (obj.attrs.test(Attrs::Optional))
+    // TODO: add attribute to mlir argument (needs PR579 attribute handling)
```

Comment:
Should this be emitting a warning or something?

Reply:
I do not think it is worth it. Right now the mlir attribute would have no functional purpose. It could be used as a hint to the optimizer that this fir.ref may be null, while others dummy fir.ref should never be null.
So it is more a TODO intended to propagate the maximum of information into FIR, rather than something that we would do wrong.